src/user-status/userStatusReducers.js

```diff
@@ -0,0 +1,41 @@
+/* @flow strict-local */
+import type { UserStatusState, Action } from '../types';
+import {
+  LOGOUT,
+  LOGIN_SUCCESS,
+  ACCOUNT_SWITCH,
+  REALM_INIT,
+  EVENT_USER_STATUS_UPDATE,
+} from '../actionConstants';
+import { NULL_OBJECT } from '../nullObjects';
+
+const initialState: UserStatusState = NULL_OBJECT;
+
+export default (state: UserStatusState = initialState, action: Action): UserStatusState => {
+  switch (action.type) {
+    case LOGOUT:
+    case LOGIN_SUCCESS:
+    case ACCOUNT_SWITCH:
+      return initialState;
+
+    case REALM_INIT:
+      return action.data.user_status || initialState;
+
+    case EVENT_USER_STATUS_UPDATE: {
+      const newUserStatus = { ...state[action.user_id] } || {};
+      if (action.away !== undefined) {
+        newUserStatus.away = action.away;
+      }
+      if (action.status_text !== undefined) {
+        newUserStatus.status_text = action.status_text;
+      }
+      return {
+        ...state,
+        [action.user_id]: newUserStatus,
+      };
```

Comment:
Hmm, this is all kind of awkwardly verbose.

How about we let the values in our state be `undefined` rather than absent? Then we can write something like
```
const status = state[action.user_id] || {};
return {
  ...state,
  [action.user_id]: {
    away: action.away !== undefined ? action.away : status.away,
    status_text: action.status_text !== undefined ? action.status_text : status.status_text,
  },
};
```

Reply:
(I don't love that version either, but it feels a little more compact and direct.)

Reply:
This code has some drawbacks.
If previously we hadn't set `away` or `status_text` it will now be set to `undefined`.

I am going in the other direction. The value `away` will never be `false` but will miss completely, `status_text` will never be `null`, `undefined` or an empty string with length=0 but will also be deleted.

Reply:
> If previously we hadn't set `away` or `status_text` it will now be set to `undefined`.

Well yes, that is what I said :slightly_smiling_face: :

> How about we let the values in our state be `undefined` rather than absent?

More precisely, that would mean that each of those properties would always be set, whether to `undefined` or to something else.


> I am going in the other direction. The value `away` will never be `false` but will miss completely, `status_text` will never be `null`, `undefined` or an empty string with length=0 but will also be deleted.

OK, cool -- I'll look at the resulting code shortly.

Reply:
OK, now that I see the intended semantics for these (`away` has just two states, and `status_text` as empty-string is synonymous with missing), I think probably the cleanest thing would actually be to have a type like this in our Redux state tree: `{| away: boolean, status_text: string |}`. Then this code would look like
```
/** The status a user implicitly has when not present in the state. */
export const DEFAULT_STATUS = { away: false, status_text: '' };
// ...

const status = state[action.user_id] || DEFAULT_STATUS;
return {
  ...state,
  [action.user_id]: {
    away: action.away !== undefined ? action.away : status.away,
    status_text: action.status_text !== undefined ? action.status_text : status.status_text,
  },
};
```

One advantage of this would be that it makes explicit what it means when we don't have this kind of data on a given user -- unlike for many other kinds of data where it would mean "we don't know", here it means "they have these default values". The `export` is there for a selector to use to encapsulate that fact.

I'm OK merging a version with the existing design, though.